### PR TITLE
timely-util: separate range bounds from `Partitioned`

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

This PR is motivated by the MySQL source which will have a nicer time defining its timestamp with this simplification in place.

Previously the `Partitioned` timestamp type was opinionated about the structure of the range of partitions, enforcing the existence of positive and negative infinity elements (even if the inner type had a perfectly fine minimum and maximum value) and the distinction between singleton inclusive sets from range exclusive sets.

This PR simplifies the `Partitioned` timestamp to always encode an inclusive range of elements made up of some type `P` that must have extremas defined. The previous semantics of additional elements can then be regained by using an appropriate `P` type and this is exactly what the Kafka source timestamp has been changed to do.

### Tips for reviewer

The first commit is the one that needs close reviewing since that's the one changing the API. The second commit is a mechanical change of all usages.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/source/kafka.rs [hunk 1]
```diff
@@ -126,15 +126,15 @@ pub struct KafkaOffsetCommiter {
 impl SourceRender for KafkaSourceConnection {
     type Key = Option<Vec<u8>>;
     type Value = Option<Vec<u8>>;
-    type Time = Partitioned<PartitionId, MzOffset>;
+    type Time = Partitioned<RangeBound<PartitionId>, MzOffset>;
 
     const STATUS_NAMESPACE: StatusNamespace = StatusNamespace::Kafka;
 
-    fn render<G: Scope<Timestamp = Partitioned<PartitionId, MzOffset>>>(
+    fn render<G: Scope<Timestamp = Partitioned<RangeBound<PartitionId>, MzOffset>>>(
```
Comment (sploiselle) [inline src/storage/src/source/kafka.rs:133]: A note here (or elsewhere) about why kafka uses rangebounds (awaiting refactor) would be helpful context if we both get hit by buses.

**Review by sploiselle — APPROVED** [review]
Looks just as expected––such good insight to simplify this.

File: src/storage/src/source/kafka.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/source/kafka.rs:137]: done, thanks for the quick review!

File: src/storage/src/source/kafka.rs [hunk 2]
```diff
@@ -126,15 +126,19 @@ pub struct KafkaOffsetCommiter {
 impl SourceRender for KafkaSourceConnection {
     type Key = Option<Vec<u8>>;
     type Value = Option<Vec<u8>>;
-    type Time = Partitioned<PartitionId, MzOffset>;
+    // TODO(petrosagg): The type used for the partition (RangeBound<PartitionId>) doesn't need to
+    // be so complicated and we could instead use `Partitioned<PartitionId, Option<u64>>` where all
+    // ranges are inclusive and a time of `None` signifies that a particular partition is not
+    // present. This requires an shard migration of the remap shard.
```
Comment (guswynn) [inline src/storage/src/source/kafka.rs:132]: ah this makes sense!

File: src/storage/src/source/kafka.rs (hunk 1 above)
Comment (guswynn) [reply src/storage/src/source/kafka.rs:137]: thank you!!!

File: src/timely-util/src/order.rs [hunk 3]
```diff
@@ -16,140 +16,87 @@
...
-/// Laying out the order in a graph we'd see something like this:
-///
-/// ```text
-///           -----(P1, T0)---(P1, T1)--- ... ---(P1, Tn)
-///          / ----(P2, T0)---(P2, T1)--- ... ---(P2, Tn)
-///         / / ,--(P3, T0)---(P3, T1)--- ... ---(P3, Tn)
-/// Bottom-+-+-+     .
-///         \        .
-///          \       .
-///           '----(Pn, T0)---(Pn, T1)--- ... ---(Pn, Tn)
```
Comment (guswynn) [inline src/timely-util/src/order.rs:54]: ahhhh

i liked this doc comment!

**Pull request merged**